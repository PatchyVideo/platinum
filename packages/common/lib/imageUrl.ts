export interface CoverImageMeta {
  image?: string | null
}

export function getCoverImage(meta: CoverImageMeta): string {
  if (meta.image && meta.image !== 'default') return 'https://patchyvideo.com/be/images/covers/' + meta.image
  return 'https://patchyvideo.com/be/images/covers/default'
}
export function getSiteImage(name: string): string {
  return new URL(`../assets/WebIcons/${name}.png`, import.meta.url).toString()
}
