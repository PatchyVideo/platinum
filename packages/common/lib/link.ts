export function openInNewTab(url: string) {
  window.open(url, '_blank')
}
