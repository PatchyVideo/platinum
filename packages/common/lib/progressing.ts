import i18n from '@/locales/index'
export function progressing(fun = i18n.global.t('others.WIP1')): void {
  alert(fun + i18n.global.t('others.WIP2'))
}
