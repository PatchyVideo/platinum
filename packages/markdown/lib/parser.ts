import MarkdownIt from 'markdown-it'
import { Parser as HTMLParser } from 'htmlparser2'
import { encode as encodeHTML } from 'html-entities'
import hljs from 'highlight.js/lib/common'
import tlds from 'tlds'

const markdownIt = new MarkdownIt({
  html: true,
  breaks: true,
  linkify: true,
  highlight: function (str, language) {
    if (language && hljs.getLanguage(language)) {
      try {
        return hljs.highlight(str, { language }).value
      } catch (__) {
        // if error, just run as not renderd.
      }
    }
    return '' // use external default escaping
  },
})

function match(reg: RegExp, text: string): number {
  if (reg.test(text)) {
    const m = text.match(reg)
    return m ? m[0].length : 0
  }
  return 0
}

function val(reg: RegExp) {
  return (text: string, pos: number) => {
    const tail = text.slice(pos)
    return match(reg, tail)
  }
}

markdownIt.linkify
  .set({
    fuzzyLink: true,
    fuzzyEmail: false,
    fuzzyIP: false,
  })
  .tlds(tlds) // full tld list
  .tlds('onion', true)
  .add('ac', {
    validate: val(/\d+/),
    normalize(match) {
      match.url = 'https://www.acfun.cn/v/' + match.url
    },
  })
  .add('av', {
    validate: val(/\d+/),
    normalize(match) {
      match.url = 'https://www.bilibili.com/video/' + match.url
    },
  })
  .add('bv', {
    validate: val(/[a-zA-Z0-9]+/),
    normalize(match) {
      match.url = 'https://www.bilibili.com/video/' + match.url
    },
  })
  .add('sm', {
    validate: val(/\d+/),
    normalize(match) {
      match.url = 'https://www.nicovideo.jp/watch/' + match.url
    },
  })
  .add('youtube-', {
    validate: val(/[-\w]+/),
    normalize(match) {
      match.url = 'https://www.youtube.com/watch?v=' + match.url.replace(/^youtube-/, '')
    },
  })
  .add('mylist/', {
    validate: val(/\d+/),
    normalize(match) {
      match.url = 'https://www.nicovideo.jp/' + match.url
    },
  })

interface MarkdownItEnv {
  last: {
    level: number
    url: URL
  }[]
}

const DRlink_open =
  markdownIt.renderer.rules.link_open ||
  function (tokens, idx, options, env, self) {
    return self.renderToken(tokens, idx, options)
  }

markdownIt.renderer.rules.link_open = function (tokens, idx, options, env: MarkdownItEnv, self) {
  const link = tokens[idx].attrGet('href')
  if (link && !(tokens[idx + 1].type === 'text' && tokens[idx + 1].content.trim() === link))
    env.last.push({
      level: tokens[idx].level,
      url: new URL(link),
    })
  return DRlink_open(tokens, idx, options, env, self)
}

const DRlink_close =
  markdownIt.renderer.rules.link_close ||
  function (tokens, idx, options, env, self) {
    return self.renderToken(tokens, idx, options)
  }

markdownIt.renderer.rules.link_close = function (tokens, idx, options, env: MarkdownItEnv, self) {
  return (
    env.last.reduce((pv: string | undefined, cv, index) => {
      if (tokens[idx].level === cv.level) {
        const str = `<span class="text-xs text-gray-600 dark:text-gray-300">[${cv.url.hostname.replace(
          /^www./,
          ''
        )}]</span>`
        env.last.splice(index, 1)
        return DRlink_close(tokens, idx, options, env, self) + str
      }
      return pv
    }, undefined) || DRlink_close(tokens, idx, options, env, self)
  )
}

/* face */
// { name: url }
const faces: Record<string, string> = {}
markdownIt.inline.ruler.before('link', 'face', (state, silent) => {
  if (silent) return false

  if (state.posMax - state.pos < 10) return false
  if (!state.src.slice(state.pos).startsWith('[[face:')) return false

  const start = state.pos
  const max = state.posMax
  state.pos++

  let found
  while (state.pos < state.posMax) {
    if (state.src.charCodeAt(state.pos) === 0x5d /* ] */) {
      found = true
      break
    }
    state.md.inline.skipToken(state)
  }

  if (!found || start + 1 === state.pos) {
    state.pos = start
    return false
  }

  const face = markdownIt.utils.unescapeMd(state.src.slice(start + 7, state.pos))

  if (face.match(/\s/)) {
    state.pos = start
    return false
  }

  state.posMax = state.pos
  state.pos = start + 1

  if (face in faces) {
    const token = state.push('face', '', 0)
    token.attrSet('href', faces['face'])
    token.content = face
  } else {
    const token = state.push('text', '', 0)
    token.content = `[${face}]`
  }

  state.pos = state.posMax + 2
  state.posMax = max

  return true
})
markdownIt.renderer.rules.face = function (tokens, idx) {
  const token = tokens[idx]
  return `<img class="face inline-block w-4 h-4" src="${token.attrGet('src')}" alt="${token.content}" />`
}

const allowedTags = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'h7',
  'h8',
  'br',
  'b',
  'i',
  'strong',
  'em',
  'a',
  'pre',
  'code',
  'img',
  'tt',
  'div',
  'ins',
  'del',
  'sup',
  'sub',
  'p',
  'ol',
  'ul',
  'table',
  'thead',
  'tbody',
  'tfoot',
  'blockquote',
  'dl',
  'dt',
  'dd',
  'kbd',
  'q',
  'samp',
  'var',
  'hr',
  'ruby',
  'rt',
  'rp',
  'li',
  'tr',
  'td',
  'th',
  's',
  'strike',
  'summary',
  'details',
  'caption',
  'figure',
  'figcaption',
  'abbr',
  'bdo',
  'cite',
  'dfn',
  'mark',
  'small',
  'span',
  'time',
  'wbr',
]
const allowedAttributes = [
  'abbr',
  'accept',
  'accept-charset',
  'accesskey',
  'action',
  'align',
  'alt',
  'aria-describedby',
  'aria-hidden',
  'aria-label',
  'aria-labelledby',
  'axis',
  'border',
  'cellpadding',
  'cellspacing',
  'char',
  'charoff',
  'charset',
  'checked',
  'clear',
  'cols',
  'colspan',
  'color',
  'compact',
  'coords',
  'datetime',
  'dir',
  'disabled',
  'enctype',
  'for',
  'frame',
  'headers',
  'height',
  'hreflang',
  'hspace',
  'ismap',
  'label',
  'lang',
  'maxlength',
  'media',
  'method',
  'multiple',
  'name',
  'nohref',
  'noshade',
  'nowrap',
  'open',
  'progress',
  'prompt',
  'readonly',
  'rel',
  'rev',
  'role',
  'rows',
  'rowspan',
  'rules',
  'scope',
  'selected',
  'shape',
  'size',
  'span',
  'start',
  'summary',
  'tabindex',
  'target',
  'title',
  'type',
  'usemap',
  'valign',
  'value',
  'vspace',
  'width',
  'itemprop',
]
const voidElements = [
  'area',
  'base',
  'basefont',
  'br',
  'col',
  'command',
  'embed',
  'frame',
  'hr',
  'img',
  'input',
  'isindex',
  'keygen',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr',
]

function sanitizeHTML(src: string) {
  const stack: string[] = []
  let result = ''
  const htmlParser = new HTMLParser({
    onopentag(tagName, attribs) {
      if (/^(https?|ftps?|mailto|xmpp|ircs?):$/.test(tagName) && Object.keys(attribs).length === 1) {
        result += `<${tagName}//${Object.keys(attribs).join(' ')}>`
        return
      } else if (!allowedTags.includes(tagName)) {
        result += encodeHTML(
          '<' +
            tagName +
            Object.entries(attribs)
              .map(([k, v]) => ` ${k}="${v.replaceAll(/"/g, '&quot;')}"`)
              .join('') +
            '>'
        )
        return
      } else {
        result += '<' + tagName
        Object.entries(attribs).forEach(([attrName, attrValue]) => {
          const addAttr = (encode = false) => {
            const v = encode ? attrValue.replaceAll(/"/g, '&quot;') : attrValue
            result += attrValue ? ` ${attrName}="${v}"` : ` ${attrName}`
          }
          try {
            switch (tagName) {
              case 'a': {
                switch (attrName) {
                  case 'href':
                    if (new URL(attrValue).protocol.match(/^(https?|ftps?|mailto|xmpp|ircs?):$/)) addAttr(true)
                    break
                }
                break
              }
              case 'img': {
                switch (attrName) {
                  case 'src':
                  case 'longdesc':
                    if (new URL(attrValue).protocol.match(/^(https?):$/)) addAttr(true)
                    break
                }
                break
              }
              case 'div': {
                switch (attrName) {
                  case 'itemscope':
                  case 'itemtype':
                    addAttr()
                }
                break
              }
              case 'blockquote':
              case 'del':
              case 'ins':
              case 'q': {
                switch (attrName) {
                  case 'cite':
                    if (new URL(attrValue).protocol.match(/^(https?):$/)) addAttr(true)
                    break
                }
                break
              }
            }
            if (allowedAttributes.includes(attrName)) addAttr(true)
          } catch (e) {
            // do nothing
          }
        })
        if (voidElements.includes(tagName)) {
          result += ' />'
        } else {
          result += '>'
          stack.push(tagName)
        }
      }
    },
    ontext(data: string) {
      result += data
    },
    onclosetag(tagName) {
      const last = stack[stack.length - 1]
      if (tagName === last) {
        result += `</${tagName}>`
        stack.pop()
      } else if (!/^(https?|ftps?|mailto|xmpp|ircs?):$/.test(tagName) && !voidElements.includes(tagName)) {
        result += encodeHTML(`</${tagName}>`)
      }
    },
  })
  htmlParser.write(src)
  htmlParser.end()
  return result
}

export function render(src: string): string {
  let res
  try {
    res = markdownIt.render(sanitizeHTML(src).replace(/\[\[表情:(\p{L}+)\]\]/gu, '[[face:$1]]'), { last: [] })
  } catch (e) {
    res = `Error throwed from Markdown parser:\n${e.name}: ${e.message}`
  }
  return res
}
