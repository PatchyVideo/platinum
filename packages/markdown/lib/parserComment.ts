import MarkdownIt from 'markdown-it'

import { facePlugin } from './parserCommon'

const markdownIt = new MarkdownIt({
  html: false,
  breaks: true,
  linkify: true,
}).use(facePlugin)

markdownIt.block.ruler.enableOnly(['paragraph'])

export function render(src: string): string {
  let res
  try {
    res = markdownIt.render(src.replace(/\[\[表情:(\p{L}+)\]\]/gu, '[[face' /* escape unocss */ + ':$1]]'), {
      last: [],
    })
  } catch (e) {
    console.error(e)
    res = `<div class="font-mono">Error throwed from Markdown parser: ${e}<br />${
      e instanceof Error && e.stack ? e.stack.replaceAll('\n', '<br />') : String(e)
    }</div>`
  }
  return res
}
