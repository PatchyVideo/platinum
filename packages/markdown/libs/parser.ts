import MarkdownIt from 'markdown-it'
import { Parser as HTMLParser } from 'htmlparser2'
import { encode as encodeHTML } from 'html-entities'
import hljs from 'highlight.js/lib/common'
import { facePlugin, linkifyAdditionPlugin } from './parserCommon'

const markdownIt = new MarkdownIt({
  html: true,
  breaks: true,
  linkify: true,
  highlight(str, language) {
    if (language && hljs.getLanguage(language)) {
      try {
        return hljs.highlight(str, { language }).value
      }
      catch (__) {
        // if error, just run as not renderd.
      }
    }
    return '' // use external default escaping
  },
})
  .use(linkifyAdditionPlugin)
  .use(facePlugin)

interface MarkdownItEnv {
  last: {
    level: number
    url: URL
  }[]
}

const DRlink_open
  = markdownIt.renderer.rules.link_open
  || function (tokens, idx, options, env, self) {
    return self.renderToken(tokens, idx, options)
  }

markdownIt.renderer.rules.link_open = function (tokens, idx, options, env: MarkdownItEnv, self) {
  const link = tokens[idx].attrGet('href')
  if (link && !(tokens[idx + 1].type === 'text' && tokens[idx + 1].content.trim() === link)) {
    env.last.push({
      level: tokens[idx].level,
      url: new URL(link),
    })
  }
  return DRlink_open(tokens, idx, options, env, self)
}

const DRlink_close
  = markdownIt.renderer.rules.link_close
  || function (tokens, idx, options, env, self) {
    return self.renderToken(tokens, idx, options)
  }

markdownIt.renderer.rules.link_close = function (tokens, idx, options, env: MarkdownItEnv, self) {
  return (
    env.last.reduce((pv: string | undefined, cv, index) => {
      if (tokens[idx].level === cv.level) {
        const str = `<span class="text-xs text-gray-600 dark:text-gray-300">[${cv.url.hostname.replace(
          /^www./,
          '',
        )}]</span>`
        env.last.splice(index, 1)
        return DRlink_close(tokens, idx, options, env, self) + str
      }
      return pv
    }, undefined) || DRlink_close(tokens, idx, options, env, self)
  )
}

const allowedTags = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'h7',
  'h8',
  'br',
  'b',
  'i',
  'strong',
  'em',
  'a',
  'pre',
  'code',
  'img',
  'tt',
  'div',
  'ins',
  'del',
  'sup',
  'sub',
  'p',
  'ol',
  'ul',
  'table',
  'thead',
  'tbody',
  'tfoot',
  'blockquote',
  'dl',
  'dt',
  'dd',
  'kbd',
  'q',
  'samp',
  'var',
  'hr',
  'ruby',
  'rt',
  'rp',
  'li',
  'tr',
  'td',
  'th',
  's',
  'strike',
  'summary',
  'details',
  'caption',
  'figure',
  'figcaption',
  'abbr',
  'bdo',
  'cite',
  'dfn',
  'mark',
  'small',
  'span',
  'time',
  'wbr',
]
const allowedAttributes = [
  'abbr',
  'accept',
  'accept-charset',
  'accesskey',
  'action',
  'align',
  'alt',
  'aria-describedby',
  'aria-hidden',
  'aria-label',
  'aria-labelledby',
  'axis',
  'border',
  'cellpadding',
  'cellspacing',
  'char',
  'charoff',
  'charset',
  'checked',
  'clear',
  'cols',
  'colspan',
  'color',
  'compact',
  'coords',
  'datetime',
  'dir',
  'disabled',
  'enctype',
  'for',
  'frame',
  'headers',
  'height',
  'hreflang',
  'hspace',
  'ismap',
  'label',
  'lang',
  'maxlength',
  'media',
  'method',
  'multiple',
  'name',
  'nohref',
  'noshade',
  'nowrap',
  'open',
  'progress',
  'prompt',
  'readonly',
  'rel',
  'rev',
  'role',
  'rows',
  'rowspan',
  'rules',
  'scope',
  'selected',
  'shape',
  'size',
  'span',
  'start',
  'summary',
  'tabindex',
  'target',
  'title',
  'type',
  'usemap',
  'valign',
  'value',
  'vspace',
  'width',
  'itemprop',
]
const voidElements = [
  'area',
  'base',
  'basefont',
  'br',
  'col',
  'command',
  'embed',
  'frame',
  'hr',
  'img',
  'input',
  'isindex',
  'keygen',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr',
]

function sanitizeHTML(src: string) {
  const stack: string[] = []
  let result = ''
  const htmlParser = new HTMLParser({
    onopentag(tagName, attribs) {
      if (/^(https?|ftps?|mailto|xmpp|ircs?):$/.test(tagName) && Object.keys(attribs).length === 1) {
        result += `<${tagName}//${Object.keys(attribs).join(' ')}>`
      }
      else if (!allowedTags.includes(tagName.toLowerCase())) {
        result += encodeHTML(
          `<${
            tagName
            }${Object.entries(attribs)
              .map(([k, v]) => ` ${k}="${v.replaceAll(/"/g, '&quot;')}"`)
              .join('')
            }>`,
        )
      }
      else {
        result += `<${tagName}`
        Object.entries(attribs).forEach(([attrName, attrValue]) => {
          const addAttr = (encode = false) => {
            const v = encode ? attrValue.replaceAll(/"/g, '&quot;') : attrValue
            result += attrValue ? ` ${attrName}="${v}"` : ` ${attrName}`
          }
          try {
            switch (tagName) {
              case 'a': {
                switch (attrName) {
                  case 'href':
                    if (new URL(attrValue).protocol.match(/^(https?|ftps?|mailto|xmpp|ircs?):$/))
                      addAttr(true)
                    break
                }
                break
              }
              case 'img': {
                switch (attrName) {
                  case 'src':
                  case 'longdesc':
                    if (new URL(attrValue).protocol.match(/^(https?):$/))
                      addAttr(true)
                    break
                }
                break
              }
              case 'div': {
                switch (attrName) {
                  case 'itemscope':
                  case 'itemtype':
                    addAttr()
                }
                break
              }
              case 'blockquote':
              case 'del':
              case 'ins':
              case 'q': {
                switch (attrName) {
                  case 'cite':
                    if (new URL(attrValue).protocol.match(/^(https?):$/))
                      addAttr(true)
                    break
                }
                break
              }
            }
            if (allowedAttributes.includes(attrName.toLowerCase()))
              addAttr(true)
          }
          catch (e) {
            // do nothing
          }
        })
        if (voidElements.includes(tagName.toLowerCase())) {
          result += ' />'
        }
        else {
          result += '>'
          stack.push(tagName)
        }
      }
    },
    ontext(data: string) {
      result += data
      // console.log(data)
    },
    onclosetag(tagName) {
      const last = stack[stack.length - 1]
      if (tagName === last) {
        result += `</${tagName}>`
        stack.pop()
      }
      else if (
        !/^(https?|ftps?|mailto|xmpp|ircs?):$/.test(tagName)
        && !voidElements.includes(tagName.toLowerCase())
      ) {
        result += encodeHTML(`</${tagName}>`)
      }
    },
  })
  htmlParser.write(src)
  htmlParser.end()
  return result
}

export function render(src: string): string {
  let res
  try {
    res = markdownIt.render(
      sanitizeHTML(src).replace(/\[\[表情:(\p{L}+)\]\]/gu, '[[face' /* escape unocss */ + ':$1]]'),
      { last: [] },
    )
  }
  catch (e) {
    console.error(e)
    res = `<div class="font-mono">Error throwed from Markdown parser: ${e}<br />${
      e instanceof Error && e.stack ? e.stack.replaceAll('\n', '<br />') : String(e)
    }</div>`
  }
  return res
}
