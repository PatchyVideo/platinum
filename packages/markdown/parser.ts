import MarkdownIt from 'markdown-it'
import hljs from 'highlight.js'
import tlds from 'tlds'

const markdownIt = new MarkdownIt({
  breaks: true,
  linkify: true,
  highlight: function (str, lang) {
    if (lang && hljs.getLanguage(lang)) {
      try {
        return hljs.highlight(lang, str).value
      } catch (__) {
        // if error, just run as not renderd.
      }
    }
    return '' // use external default escaping
  },
})

function match(reg: RegExp, text: string): number {
  if (reg.test(text)) {
    const m = text.match(reg)
    return m ? m[0].length : 0
  }
  return 0
}

markdownIt.linkify
  .tlds(tlds) // full tld list
  .tlds('onion', true)
  .add('git:', 'http:')
  .add('ac', {
    validate(text, pos, self) {
      const tail = text.slice(pos)
      if (!self.re.aclink) self.re.aclink = /\d+/
      return match(self.re.aclink, tail)
    },
    normalize(match) {
      match.url = 'https://www.acfun.cn/v/' + match.url
    },
  })
  .add('av', {
    validate(text, pos, self) {
      const tail = text.slice(pos)
      if (!self.re.avlink) self.re.avlink = /\d+/
      return match(self.re.avlink, tail)
    },
    normalize(match) {
      match.url = 'https://www.bilibili.com/video/' + match.url
    },
  })
  .add('bv', {
    validate(text, pos, self) {
      const tail = text.slice(pos)
      if (!self.re.bvlink) self.re.bvlink = /[a-zA-Z0-9]+/
      return match(self.re.bvlink, tail)
    },
    normalize(match) {
      match.url = 'https://www.bilibili.com/video/' + match.url
    },
  })
  .add('sm', {
    validate(text, pos, self) {
      const tail = text.slice(pos)
      if (!self.re.smlink) self.re.smlink = /\d+/
      return match(self.re.smlink, tail)
    },
    normalize(match) {
      match.url = 'https://www.nicovideo.jp/watch/' + match.url
    },
  })
  .add('youtube-', {
    validate(text, pos, self) {
      const tail = text.slice(pos)
      if (!self.re.u2blink) self.re.u2blink = /[-\\w]+/
      return match(self.re.u2blink, tail)
    },
    normalize(match) {
      match.url = 'https://www.youtube.com/watch?v=' + match.url.replace(/^youtube-/, '')
    },
  })
  .add('mylist', {
    validate(text, pos, self) {
      const tail = text.slice(pos)
      if (!self.re.ncllink) self.re.ncllink = /[-\\w]+/
      return match(self.re.ncllink, tail)
    },
    normalize(match) {
      match.url = 'https://www.nicovideo.jp/' + match.url
    },
  })

export default markdownIt
