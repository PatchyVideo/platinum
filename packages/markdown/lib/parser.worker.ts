/// <reference lib="webworker" />
import { render } from './parser'

const ctx: DedicatedWorkerGlobalScope = (self as unknown) as DedicatedWorkerGlobalScope

ctx.onmessage = (e) => {
  ctx.postMessage({ id: e.data.id, raw: e.data.text, text: render(e.data.text) })
}
