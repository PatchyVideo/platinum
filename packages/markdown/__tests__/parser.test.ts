import { describe, expect, it } from 'vitest'
import { render } from '../lib/parser'

describe('Markdown Parser', () => {
  it('can render plain text', () => {
    const result = render('Hello World!')

    expect(result).toBe('<p>Hello World!</p>\n')
  })
  it('can render markdown', () => {
    const result = render('# Hello World!')

    expect(result).toBe('<h1>Hello World!</h1>\n')
  })
  it('can render allowed dom', () => {
    const result = render(`
plain text
<h1>h1</h1>
<br />
<a href="https://example.com">a</a>
<img src="https://example.com/example.png" />`)

    expect(result).toMatchSnapshot()
  })
  it('can filter xss dom', () => {
    const result = render(`
<script>alert(1)</script>
<style>body{color:red}</style>
<img src=x onerror=alert(1)>`)

    expect(result).toMatchSnapshot()
  })
  it('can add hint to link', () => {
    const result = render('[Example](https://example.com/foo)')

    expect(result).toMatchSnapshot()
  })
  it('can render pure link', () => {
    const result = render(`
https://example.com
av926598384
BV11T4y1j7ni
ac17966946
sm38945360
youtube-zOmbs-UMwx0
mylist/65545020`)

    expect(result).toMatchSnapshot()
  })
  it('should not filter <link>', () => {
    const result = render('<https://example.com>')

    expect(result).toMatchSnapshot()
  })
})
