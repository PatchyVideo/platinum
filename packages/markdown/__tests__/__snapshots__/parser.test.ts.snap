// Vitest Snapshot v1

exports[`Markdown Parser > can add hint to link 1`] = `
"<p><a href=\\"https://example.com/foo\\">Example</a><span class=\\"text-xs text-gray-600 dark:text-gray-300\\">[example.com]</span></p>
"
`;

exports[`Markdown Parser > can filter xss dom 1`] = `
"<p>&lt;script&gt;alert(1)&lt;/script&gt;<br>
&lt;style&gt;body{color:red}&lt;/style&gt;<br>
<img /></p>
"
`;

exports[`Markdown Parser > can render allowed dom 1`] = `
"<p>plain text</p>
<h1>h1</h1>
<br />
<a href=\\"https://example.com\\">a</a>
<img src=\\"https://example.com/example.png\\" />"
`;

exports[`Markdown Parser > can render pure link 1`] = `
"<p><a href=\\"https://example.com\\">https://example.com</a><br>
<a href=\\"https://www.bilibili.com/video/av926598384\\">av926598384</a><span class=\\"text-xs text-gray-600 dark:text-gray-300\\">[bilibili.com]</span><br>
<a href=\\"https://www.bilibili.com/video/BV11T4y1j7ni\\">BV11T4y1j7ni</a><span class=\\"text-xs text-gray-600 dark:text-gray-300\\">[bilibili.com]</span><br>
<a href=\\"https://www.acfun.cn/v/ac17966946\\">ac17966946</a><span class=\\"text-xs text-gray-600 dark:text-gray-300\\">[acfun.cn]</span><br>
<a href=\\"https://www.nicovideo.jp/watch/sm38945360\\">sm38945360</a><span class=\\"text-xs text-gray-600 dark:text-gray-300\\">[nicovideo.jp]</span><br>
<a href=\\"https://www.youtube.com/watch?v=zOmbs-UMwx0\\">youtube-zOmbs-UMwx0</a><span class=\\"text-xs text-gray-600 dark:text-gray-300\\">[youtube.com]</span><br>
<a href=\\"https://www.nicovideo.jp/mylist/65545020\\">mylist/65545020</a><span class=\\"text-xs text-gray-600 dark:text-gray-300\\">[nicovideo.jp]</span></p>
"
`;

exports[`Markdown Parser > should not filter <link> 1`] = `
"<p><a href=\\"https://example.com\\">https://example.com</a></p>
"
`;
